docs/function_guide.rst

```diff
@@ -0,0 +1,264 @@
+==================================
+Writing libEnsemble User Functions
+==================================
+
+libEnsemble coordinates ensembles of calculations performed by three main
+functions: a :ref:`Generator Function<api_gen_f>`, a :ref:`Simulator Function<api_sim_f>`,
+and an :ref:`Allocation Functions<api_alloc_f>`, or ``gen_f``, ``sim_f``, and
+``alloc_f`` respectively. These are all referred to as User Functions. Although
+libEnsemble includes several ready-to-use User Functions like
+:doc:`APOSMM<examples/aposmm>`, it's expected that most users will write their own.
+This guide serves as an overview of both necessary and optional components for
+writing different kinds of User Functions, and common development patterns.
+
+Generator Functions
+===================
+
+As described in the :ref:`API<api_gen_f>`, the ``gen_f`` is called by a
+libEnsemble worker via the following::
+
+    out = gen_f(H[gen_specs['in']][sim_ids_from_allocf], persis_info, gen_specs, libE_info)
+
```

Comment:
Somewhere here might mention the case of persistent gens, where we currently tend to have gen_specs empty and let the allocation function determine what is sent (though we might want to change this).

Reply:
Agree on mentioning the difference between the default gens (which only have whatever was given to/from H to know about the past gen/sim work) and persistent gens (which occupy a worker for a period of time and do not need to be re-initialized every time their past work is completed and returned from the manager).